nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/CommonClassLoaderServiceImpl.java

```diff
@@ -189,87 +178,57 @@ private static String urlsToClassPath(Stream<URL> urls) {
         return urls.map(FileUtils::toFile).map(File::getAbsolutePath).collect(Collectors.joining(File.pathSeparator));
     }
 
-    private List<File> findDerbyClient() {
-        final String DERBY_HOME_PROP = "AS_DERBY_INSTALL";
-        StartupContext startupContext = env.getStartupContext();
-        Properties arguments = null;
+    private List<File> findDerbyJars(File installDir) {
+        Path derbyHome = getDerbyDir(installDir);
+        LOG.log(CONFIG, "Using derby home: {0}", derbyHome);
 
-        if (startupContext != null) {
-            arguments = startupContext.getArguments();
+        final File derbyLib = derbyHome.resolve("lib").toFile();
+        if (!derbyLib.exists()) {
+            LOG.info(KernelLoggerInfo.cantFindDerby);
+            return Collections.emptyList();
         }
 
-        String derbyHome = null;
-
-        if (arguments != null) {
-            derbyHome = arguments.getProperty(DERBY_HOME_PROP,
-                System.getProperty(DERBY_HOME_PROP));
-        }
+        return Arrays
+            .asList(derbyLib.listFiles((dir, name) -> name.endsWith(".jar") && !name.startsWith("derbyLocale_")));
+    }
 
-        File derbyLib = null;
-        if (derbyHome != null) {
-            derbyLib = new File(derbyHome, "lib");
+    private static Path getDerbyDir(File installDir) {
+        String derbyHomeProperty = System.getProperty(DERBY_ROOT_PROP_NAME);
+        if (derbyHomeProperty == null) {
+            return installDir.toPath().resolve(Path.of("..", "javadb"));
         }
-        if (derbyLib == null || !derbyLib.exists()) {
-            logger.info(KernelLoggerInfo.cantFindDerby);
-            return Collections.emptyList();
+        Path derbyHome = Path.of(derbyHomeProperty);
+        if (derbyHome.isAbsolute()) {
+            return derbyHome;
         }
-
-        return Arrays.asList(derbyLib.listFiles(new FilenameFilter(){
-            @Override
-            public boolean accept(File dir, String name) {
-                // Include only files having .jar extn and exclude all localisation jars, because they are
-                // already mentioned in the Class-Path header of the main jars
-                return (name.endsWith(".jar") && !name.startsWith("derbyLocale_"));
-            }
-        }));
+        return new File(installDir, "config").toPath().resolve(derbyHome);
```

Comment:
Why does this use "installDir/config" as the root path? Is it an arbitrary path introduced here if the property specifies a relative path? Should we also use it here: https://github.com/eclipse-ee4j/glassfish/blob/7bb61e595e5f43b02406a2264a7d351f2389db0e/appserver/connectors/connectors-runtime/src/main/java/com/sun/enterprise/connectors/util/DriverLoader.java#L495 ?

Reply:
Aha, yes, I added that because when it is a relative path, it is always resolved to config dir. However all these relative paths we use now are bit insecure, so this is rather temporary. And maybe not needed any more, after recent changes, I have to check.